### PR TITLE
[WIP] Worker thread

### PR DESCRIPTION
**Pull request by psie**
This is a draft of worker thread concept. Includes a use case of cleaning up after a thread that exited. Awaiting comments.

### CODE REVIEW

File: sys/thread.c [hunk 1]
```diff
@@ -105,6 +113,12 @@ noreturn void thread_exit() {
     ;
 }
 
+void thread_join(void *p) {
```
Comment (cahirwpz) [inline sys/thread.c]: Please add `condvar_t td_cvexit` to `thread_t` structure and use `cv_broadcast` to wakeup threads that have been waiting for a thread to finish. `thread_join` takes `thread_t *` argument and will `cv_wait` on this thread's `td_cvexit`.

File: include/taskqueue.h [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+#include <queue.h>
+
+typedef struct task {
+  STAILQ_ENTRY(task) tq_link;
+  void *arg;            /* function argument */
+  void (* func)(void*); /* function pointer */
+} task_t;
+
+typedef struct tq_head taskqueue_t;
```
Comment (cahirwpz) [inline include/taskqueue.h]: I had a look at NetBSD `workqueue` implementation. You've completely forgotten about synchronization issues...

```C
typedef struct taskqueue {
  mtx_t tq_mutex;
  condvar_t tq_nonempty; /* worker waits on this cv for tq_list to become non empty */
  STAILQ_HEAD(, task) tq_list;
  thread_t *tq_worker;
};
```

**Comment by cahirwpz** [comment]
Hi, is there any progress on this task?

**Comment by cahirwpz** [comment]
@czapiga Could you take this task over and finish it. We need taskqueue to be synchronized.

File: sys/thread.c (hunk 1 above)
Comment (cahirwpz) [inline sys/thread.c]: Please use `sleepq_wait` on `td_exitcode` to implement `thread_join`.

**Comment by rafalcieslak** [comment]
Moved to #216.

**Pull request closed without merging**